src/rand.rs

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright (c) 2018 Adgear
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of
+ * this software and associated documentation files (the "Software"), to deal in
+ * the Software without restriction, including without limitation the rights to
+ * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+ * the Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+//! Random number generator based of the PCG paper (http://www.pcg-random.org/paper.html).
+
+use std::u32;
+use std::time::*;
+
+pub struct Rng {
+    state: u64,
+    inc: u64,
+}
+
+impl Default for Rng {
+    fn default() -> Self {
+        match SystemTime::now().duration_since(UNIX_EPOCH) {
+            Ok(res) => Self::seed_with(res.as_secs()),
+            Err(_) => Self::seed_with(6_364_136_223_846_793_005)
+        }
+    }
+}
+
+impl Rng {
+    fn pcg32(&mut self) -> u32 {
+        let oldstate = self.state;
+        // Advance internal state
+        self.state = u64::wrapping_add(u64::wrapping_mul(oldstate, 6_364_136_223_846_793_005u64), self.inc | 1);
+        // Calculate output function (XSH RR), uses old state for max ILP
+        let xorshifted = (((oldstate >> 18) ^ oldstate) >> 27) & 0xFFFF_FFFF;
+        let rot = (oldstate >> 59) & 0xFFFF_FFFF;
+        let v = (xorshifted >> rot) | (xorshifted << (u64::wrapping_sub(0, rot) & 31));
+        v as u32
+    }
+
+    pub fn seed_with(seed: u64) -> Self {
+        // We xor the seed with a randomly chosen number to avoid ending up with
+        // a 0 state which would be bad.
+        Self {
+            state: seed ^ 0xedef_335f_00e1_70b3,
+            inc: 12345,
```

Comment:
Ask @eloraiby. I think the one used in the state comes from the paper of the algorithm.

Reply:
So a 0 state for xorshift is real bad as it can't transition away from it. That being said, I'm not quite sure whether pcg which uses an lcg and xorshift is vulnerable to this. I'd have to think about it a bit more (although a quick test could easily answer the question).